### PR TITLE
Fix race condition and Delta Operation incomplete generation

### LINKED ISSUE #3: Delta Operation incomplete generation
I found that it cannot be rebuilt properly for a specific file during use.

## How to reproduce 

It occurs in the line 310 `io.ReadAtLeast` return value of 0 and EOF in the CreateDelta function.
This is closely related to the configuration of RSync.BlockSize and the size of new files.

```go
n, err = io.ReadAtLeast(source, buffer[validTo:validTo+r.BlockSize], r.BlockSize)
```
I made a simple modification based on the ExempleRsync method in exemple_test.go to show this situation.
```go
package rsync_test

import (
	"io"
	"os"
	"strings"
	"testing"

	"github.com/minio/rsync-go"
	"github.com/stretchr/testify/assert"
)

func Test_Rync(t *testing.T) {
	oldContent := "1234"
	newContent := "123456789123456"
	blockSize := 3

	oldReader := strings.NewReader(oldContent)
	rs := &rsync.RSync{
		BlockSize: blockSize,
	}

	// here we store the whole signature in a byte slice,
	// but it could just as well be sent over a network connection for example
	sig := make([]rsync.BlockHash, 0, 10)
	writeSignature := func(bl rsync.BlockHash) error {
		sig = append(sig, bl)
		return nil
	}
	rs.CreateSignature(oldReader, writeSignature)

	var currentReader io.Reader = strings.NewReader(newContent)

	opsOut := make(chan rsync.Operation)
	writeOperation := func(op rsync.Operation) error {
		opsOut <- op
		return nil
	}

	go func() {
		defer close(opsOut)
		rs.CreateDelta(currentReader, sig, writeOperation)
	}()

	var newWriter strings.Builder
	oldReader.Seek(0, os.SEEK_SET)

	rs.ApplyDelta(&newWriter, oldReader, opsOut)

	result := newWriter.String()
	assert.Equal(t, result, newContent)
}

```

## How to solve

Comment out lines 320 through 322

```go
	n, err = io.ReadAtLeast(source, buffer[validTo:validTo+r.BlockSize], r.BlockSize)
	validTo += n
	if err != nil {
		if err != io.EOF && err != io.ErrUnexpectedEOF {
			return err
		}
		lastRun = true

		data.head = validTo
	}
	//if n == 0 {
	//	break
	//}
```

I'm not sure this raises any other questions

### PR DESCRIPTION
**Pull request by mdouchement**
Fix #3 
- The race condition along with the different sizes of buff initialization was an issue
- As stated in #3, when `n, err = io.ReadAtLeast(source, buffer[validTo:validTo+r.BlockSize], r.BlockSize)` ended on an `io.EOF`, `n` equals `0` and made the remaining data not written.

```
$ go test -run=NONE -fuzz=FuzzSync -race
fuzz: elapsed: 0s, gathering baseline coverage: 0/76 completed
fuzz: elapsed: 0s, gathering baseline coverage: 76/76 completed, now fuzzing with 16 workers
fuzz: elapsed: 3s, execs: 22875 (7624/sec), new interesting: 7 (total: 83)
fuzz: elapsed: 6s, execs: 51585 (9567/sec), new interesting: 9 (total: 85)
fuzz: elapsed: 9s, execs: 78026 (8802/sec), new interesting: 9 (total: 85)
...
fuzz: elapsed: 10m6s, execs: 6970026 (11394/sec), new interesting: 22 (total: 110)
fuzz: elapsed: 10m9s, execs: 7004121 (11363/sec), new interesting: 22 (total: 110)
fuzz: elapsed: 10m12s, execs: 7021020 (6222/sec), new interesting: 22 (total: 110)
PASS
ok      github.com/minio/rsync-go       613.028s
```

### CODE REVIEW

File: example_test.go [hunk 1]
```diff
@@ -34,13 +33,13 @@ func ExampleRsync() {
 
 	go func() {
 		defer close(opsOut)
-		rs.CreateDelta(currentReader, sig, writeOperation)
+		_ = rs.CreateDelta(currentReader, sig, writeOperation)
 	}()
 
 	var newWriter strings.Builder
-	oldReader.Seek(0, os.SEEK_SET)
+	_, _ = oldReader.Seek(0, io.SeekStart)
```
Comment (harshavardhana) [inline example_test.go]:
```suggestion
	_, _ = oldReader.Seek(0, io.SeekStart)
```

Can you make sure to only send just this change? all other changes are not needed.

File: fuzzing_test.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
+package rsync_test
```
Comment (harshavardhana) [inline fuzzing_test.go:10]: Add copyright header from the rsync.go file

File: go.mod [hunk 3]
```diff
@@ -2,6 +2,14 @@ module github.com/minio/rsync-go
 
 go 1.17
 
-require github.com/minio/highwayhash v1.0.2
+require (
+	github.com/minio/highwayhash v1.0.3
+	github.com/stretchr/testify v1.9.0
```
Comment (harshavardhana) [inline go.mod]: do not want to use testify please use simple things like go's testing package.

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: go.mod [hunk 4]
```diff
@@ -2,6 +2,6 @@ module github.com/minio/rsync-go
 
 go 1.17
```
Comment (mdouchement) [inline go.mod]: How about:
```suggestion
go 1.22
```

In the fuzzing test, `math/rand/v2` is imported which is part of go1.22 release.
`go test` / `go vet` are broken when executed on go1.17 (in the current state of this PR), it should be still importable.
Updating a Golang project version is quite easy.

On the other hand, we can downgrade to `math/rand`.

Comment (mdouchement) [reply go.mod]: https://github.com/minio/rsync-go/blob/8dc6f97e7eef301a78c9e5b827255a7c44f76eac/.github/workflows/vulncheck.yml#L17
needs to be updated too

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**